common: add variant class

UkoeHB opened the pull request:
This is a PR in my 'upstreaming seraphis_lib project', the changes here are not used anywhere yet.

- Add a `tools::variant` class with a useful interface.

jeffro256 commented:
I've gotta ask: what's the point of having a possible `boost::blank` underlying type for the variant class? That just seems like it would add a lot of edge cases where they aren't needed.

jeffro256 commented:
@UkoeHB Thanks for the fast response! I immediately realized right after I posted that that it was kind of a dumb question since I could easily think of some uses for it and deleted it lol, but you were too fast

vtnerd commented on src/common/variant.h:
```diff
@@ -0,0 +1,135 @@
+
+    /// apply a visitor to the variant
+    template <typename VisitorT>
+    typename VisitorT::result_type visit(const VisitorT &visitor) const
```
I would `&& visitor` with `std::forward<VisitorT>(visitor)` this too since `boost::apply_visitor` is doing this anyway.

UkoeHB commented on src/common/variant.h:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    /// check if two variants have the same type
+    static bool same_type(const variant<Types...> &v1, const variant<Types...> &v2)
```
Static matches the semantics - it's a function called on two variants, not a behavior of one variant.

vtnerd replied:
I forgot about ADL briefly, are you invoking this like:
```c++
same_type(val1, val2);
```
? This should work, provided no one else adds a `same_type` function to the namespace. This could be a standalone function - I don't see anything wrong with that either.

vtnerd replied:
Nevermind, I would just leave as-is unless it becomes an issue. Might be confusing for the person to change though.

UkoeHB replied:
[This](https://github.com/UkoeHB/monero/blob/6e7b72ccc3c576f2deff936f1dd4f085a9064798/src/multisig/multisig_signing_helper_utils.cpp#L789) is the only place it's used right now.

UkoeHB commented on src/common/variant.h:
```diff
@@ -0,0 +1,135 @@
+
+    /// construct from variant type
+    template <typename T>
+    variant(const T &value) : m_value{value} {}
```
My goal is a variant with assignable value semantics, so I think this shouldn't be `explicit`.

UkoeHB commented on src/common/variant.h:
```diff
@@ -0,0 +1,135 @@
+template <typename... Types>
+class variant final
+{
+    using VType = boost::variant<boost::blank, Types...>;
```
`boost::blank` is designed specifically for this use-case afaik. Plus, a variant isn't a pointer class so I think `nullptr` is a little inconsistent.

vtnerd replied:
I would continue to use `boost::blank`. An overload with `nullptr_t` (not needed now), could definitely be useful in generic code like in my serialization library. There are situations where the concepts between pointer, optional, and variant are nearly identical - assign this thing a "dead" value - but with three different techniques for assignment, requires 3 independent functions until C++17 `constexpr if` (which also confuses typical C++ devs).

Anyway, this is why I always thought `std::nullptr_t`/`nullptr` has always seemed just as appropriate though. `std::optional` has `std::nullopt_t` and `nullopt`, and the only difference is slightly more accurate type-checking but little else. Yes, `std::optional` is not _literally_ a pointer, but the intended objectives are identical and why get so fancy with unique types (because it makes writing template/generic functions _more_ complicated).

jeffro256 commented:
Question: Do you ever plan on supporting recursive variants?

UkoeHB commented:
A variant of variants? Not sure where or why you'd want to use that.

jeffro256 commented:
They’re common for representing deserialized unstructured data in memory. For example, see epee::serialization::storage_entry in contrib/epee/include/storages/portable_storage_base.h, or Value in rapidjson.

UkoeHB commented:
Is there some particular change required to support that?

UkoeHB commented:
Changes: added `noexcept` to `.index()` and `::type_index_of()`.

UkoeHB commented:
Changes: updated `is_type()` to use a cheap index check instead of `boost::strict_get()`.

jeffro256 commented:
> Is there some particular change required to support that?

Actually, probably not but the user would have to use `boost::recursive_variant_wrapper` which is sort of a leaky abstraction.

jeffro256 commented on src/common/variant.h:
```diff
@@ -0,0 +1,149 @@
+
+    /// get the type index of a requested type (compile error for invalid types) (boost::mp11 is boost 1.66.0)
+    template <typename T>
+    static int type_index_of() noexcept
```
```suggestion
    static constexpr int type_index_of() noexcept
```

jeffro256 replied:
This works as a `constexpr`, just tested it

UkoeHB commented on src/common/variant.h:
```diff
@@ -0,0 +1,135 @@
+};
+
+template <typename... Types>
+class variant final
```
`final` is part of a class's interface. I mark everything `final` that is not explicitly intended to be inherited from. In general, I feel 'implementation inheritance' should be avoided unless there is a very strong justification. Note that making it non-final would mean a virtual destructor is required.

UkoeHB commented:
Changes: add `try_unwrap()`, mark `type_index_of()` as `constexpr`.

jeffro256 commented on src/common/variant.h:
```diff
@@ -0,0 +1,156 @@
+#include <boost/none_t.hpp>
+#include <boost/variant/apply_visitor.hpp>
+#include <boost/variant/get.hpp>
+#include <boost/variant/static_visitor.hpp>
```
```suggestion
#include <boost/variant/recursive_wrapper.hpp>
#include <boost/variant/static_visitor.hpp>
```

jeffro256 commented on src/common/variant.h:
```diff
@@ -0,0 +1,156 @@
+    /// - add this to your visitor with: using variant_static_visitor::operator();
+    [[noreturn]] ResultT operator()(const boost::blank) const { variant_static_visitor_blank_err(); }
+};
+
```
```suggestion

template <typename T>
using variant_recursive_wrapper = boost::recursive_wrapper<T>;

typedef boost::recursive_variant_ variant_placeholder;

template <typename... Ts>
using make_recursive_variant = boost::make_recursive_variant<Ts...>;
```

jeffro256 reviewed (commented):
<s>This will give the variant class the ability to make recursive variants just like Boost has: https://www.boost.org/doc/libs/1_58_0/doc/html/variant/tutorial.html#variant.tutorial.recursive</s>

Don't worry about this actually. It doesn't need this to work technically, and this might be an instance of YAGNI

jeffro256 commented:
Made a PR against this PR here: https://github.com/UkoeHB/monero/pull/6. Just includes some unit tests

UkoeHB commented:
Changes: added some nice unit tests thanks to @jeffro256, added a commit on top of that with some formatting/legibility changes (will squash once @jeffro256 can take a look at the diff).

UkoeHB commented on tests/unit_tests/variant.cpp:
```diff
@@ -26,41 +26,45 @@
-#include <sstream>
-#include <type_traits>
-#include <vector>
+#include "common/variant.h"
```
Including this first means the header doesn't depend on transitive includes.

jeffro256 commented:
You cant use `this->` in static methods :)

UkoeHB commented on tests/unit_tests/variant.cpp:
```diff
@@ -26,41 +26,45 @@
 using tools::variant_static_visitor;
 
 namespace
 {
```
I still think the anonymous namespace is redundant, but leaving it in here since it's your file.

jeffro256 replied:
I particularly like using anonymous namespaces because `static` doesn't prevent ODR violations. I don't know, I don't care too much either way.

UkoeHB commented on tests/unit_tests/variant.cpp:
```diff
@@ -207,7 +219,7 @@ struct test_stringify_visitor: public variant_static_visitor<std::string>
     static void test_visitation(const Variant& v, const T& t)
     {
-        EXPECT_EQ(stringify(t), v.visit(test_stringify_visitor()));
+        EXPECT_EQ(this->stringify(t), v.visit(test_stringify_visitor()));
```
Explicit `this` helps disambiguate between free and member function calls.

UkoeHB replied:
@jeffro256 oh you're right, whoops

UkoeHB commented on tests/unit_tests/variant.cpp:
```diff
@@ -398,4 +411,5 @@ TEST(variant, ad_hoc_recursion)
-}
\ No newline at end of file
+}
+//-------------------------------------------------------------------------------------------------------------------
```
Github likes having a newline at the bottom.

vtnerd commented on src/common/variant.h:
```diff
@@ -0,0 +1,149 @@
+                        std::remove_cv_t<std::remove_reference_t<T>>,
+                        variant<Types...>
+                    >::value,
+                bool
```
I meant to delete the above comment, just whatever.
